### PR TITLE
directory_loader: Add "exclude_glob" option

### PR DESCRIPTION
**Pull request by brannt**
Add an `exclude_glob` option to exclude some files or dirs from the main glob. Sample use case: a directory structure like this
```
main_directory/
    backup/ # Don't want to load from here
    file_directory_1/
    file_directory_2/
    ...
``` 
(inspired by [LogSeq](https://logseq.com/) data structure)

### CODE REVIEW

File: libs/langchain/langchain/document_loaders/directory.py [hunk 1]
```diff
@@ -31,6 +31,7 @@ def __init__(
         self,
         path: str,
         glob: str = "**/[!.]*",
+        exclude_glob: Optional[str] = None,
```
Comment (baskaryan) [inline libs/langchain/langchain/document_loaders/directory.py:34]: maybe i don't fully understand, but should doesn't glob already have enough expressivity to exclude whatever you'd like

Comment (brannt) [reply libs/langchain/langchain/document_loaders/directory.py:34]: I don't think so, there doesn't seem to be a way to exclude anything complicated at least (for example: include all *.py files except the ones that start with test_, or, like a say here, include all subdirectories except one). This answer on SO https://stackoverflow.com/questions/20638040/glob-exclude-pattern suggests to do the exclusion in Python.

**Comment by nejch** [comment]
FYI: I implemented https://github.com/langchain-ai/langchain/pull/17316 that's consistent with the behavior in `GenericLoader` but I only just discovered this PR. But just so you're aware there's something similar in the codebase now.

**Comment by brannt** [comment]
Closed as duplicate

**Pull request closed without merging**